### PR TITLE
112 improve curriculum algorithm based on the feedback from murmurasz

### PR DESCRIPTION
**Pull request by GustawB**
Added:
Vibe-check
Global set for test eval
Fixed learning rate preservation between curriculer steps.

ROC_AUG score: https://developers.google.com/machine-learning/crash-course/classification/roc-and-auc
It seems that for our case (unbalanced classes) we should use precision/recall plots instead. Hovewer, we already have precision and recall plots (not together, but still), so I decided that for now I won't add the new plot.

### CODE REVIEW

File: src/bert_finetuning/README.md [hunk 1]
```diff
@@ -23,6 +23,10 @@ This process happens in a loop.
...
 
-## Dev notes:
-Right now, each "step" (training on new dataset) has a new trainer; I tried to reuse it, but lr was still behaving in a weird way, 
-and for instance, trainer was resuming from epoch 2.9 (saved state). I wouldn't be surprised if its fixable, but I didn't have enough time to fix it.
+## Dev notes: Sequel
+1. Fixed the learning rate preservation between the training steps in the curriculer. However, now it seems
+that at some point BERT just stops learning (learning rate is too low). More experiments needs to be conducted.
+2. Added vibe-check. After some fights, I decided to help myself with the existing functionality of the HF,
+and right now the vibecheck returns "UNKNOWN" for words that has no tokens classified as part of the coupon, 
+and "COUPON" otherwise. Because rn models aren't performing very well (low LR), most of the words have
```
Comment (SzymonKozl) [inline src/bert_finetuning/README.md:31]: Have you tried bigger lr/less aggressive shedulers?

File: src/bert_finetuning/finetuner.py [hunk 2]
```diff
@@ -194,30 +204,52 @@ def __compute_metrics(custom_labels: list, eval_preds: list) -> dict:
...
+    """
+    Function that is responsible for printing a random sample from the dataset.
 
-def train_model(model: callable, dataset: Dataset, labels: list, run_name: str, push_to_hub: bool=False, wandb_log: bool=False, curriculum_learning: bool=False, splits: int=10):
+    :param model: The model that will be used to predict the output
+    :param dataset: The dataset that contains the samples
+    """
+    classifier = pipeline("token-classification", model=model, tokenizer=__TOKENIZER, device=0, aggregation_strategy="simple")
+
+    random_index = random.randint(0, len(dataset["train"]) - 1)
```
Comment (SzymonKozl) [inline src/bert_finetuning/finetuner.py:216]: I think that for training progress visualtsation it would be nice to see evaluations across epochs on single, non-changing input.

File: fine_tuning_results/bert_second_pass.ipynb
Comment (SzymonKozl) [inline fine_tuning_results/bert_second_pass.ipynb:1]: I am suprised by the lr chart. Are we training for ~5k steps and for 4.5k steps the lr is almost zero? Is this true?

**Review by SzymonKozl — COMMENTED** [review]
gj in overall, besides from comments I have left, I would love to see output of this "vibe check" without the need to run the finetuning on my own. Maybe visualisation like that could be even used in our baczelor deezis?

File: src/bert_finetuning/README.md (hunk 1 above)
Comment (GustawB) [reply src/bert_finetuning/README.md:31]: No. I was going to, but:
1. I blew my 30$ on modal already, didnt want to waste more resources.
2. This is not scope of this task; I think this is something that should be done next week when we will perform fine-tunins.

File: fine_tuning_results/bert_second_pass.ipynb
Comment (GustawB) [reply fine_tuning_results/bert_second_pass.ipynb:1]: It makes sense; I preserve lr between runs; We start with 2e-5; after first round I think it was 1e-7; so it's not technically zero, but it pretty quickly reaches values that arent distinguishable from zero for the 64bit machine.

File: src/bert_finetuning/finetuner.py (hunk 2 above)
Comment (GustawB) [reply src/bert_finetuning/finetuner.py]: Added a parameter for specifying this.

**Comment by GustawB** [comment]
> gj in overall, besides from comments I have left, I would love to see output of this "vibe check" without the need to run the finetuning on my own. Maybe visualisation like that could be even used in our baczelor deezis?

Made this function more accessible to use outside of the training loop.

File: src/bert_finetuning/README.md (hunk 1 above)
Comment (SzymonKozl) [reply src/bert_finetuning/README.md:31]: ok fair, closing

File: fine_tuning_results/bert_second_pass.ipynb
Comment (SzymonKozl) [reply fine_tuning_results/bert_second_pass.ipynb:1]: okay, we need to counter this in our future experiments

**Comment by SzymonKozl** [comment]
Tbh not exactly what I wanted, but for now let it be. Just make sure that for the proper fine-tuning we will save the vibe-check results somewhere.

**Review by SzymonKozl — APPROVED** [review]
